sql/distinct.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tamir Duberstein (tamird@gmail.com)
+
+package sql
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/sql/parser"
+)
+
+// distinct constructs a distinctNode.
+func (*planner) distinct(n *parser.Select) *distinctNode {
+	if n.Distinct {
+		distinct := distinctNode{}
+		distinct.suffixSeen = make(map[string]struct{})
+		return &distinct
+	}
+	return nil
+}
+
+type distinctNode struct {
+	planNode
+	// All the columns that are part of the Sort. Set to nil if no-sort, or
+	// sort used an expression that was not part of the requested column set.
+	columnsInOrder []bool
+	// encoding of the columnsInOrder columns for the previous row.
+	prefixSeen string
+	// encoding of the non-columnInOrder columns for rows sharing the same
+	// prefixSeen value.
+	suffixSeen map[string]struct{}
+}
+
+// wrap the supplied planNode with the distinctNode if distinct is required.
+func (n *distinctNode) wrap(plan planNode) planNode {
+	if n == nil {
+		return plan
+	}
+	n.planNode = plan
+	if len(plan.Ordering()) > 0 {
+		n.columnsInOrder = make([]bool, len(plan.Columns()))
+	}
+	for _, p := range plan.Ordering() {
+		if p < 0 {
+			p = -p
+		}
+		if p <= len(n.columnsInOrder) {
+			n.columnsInOrder[p-1] = true
+		} else {
+			// Cannot use sort order for distinct.
+			n.columnsInOrder = nil
+			break
+		}
+	}
+	return n
+}
+
+func (n *distinctNode) Next() bool {
+	for n.planNode.Next() {
+		// Detect duplicates
+		prefix, suffix := n.encodeValues(n.Values())
+		pKey := string(prefix)
```

Comment:
Converting to a string makes the comparison easier, but forces another memory allocation. This is necessary for `sKey` since you can't use `[]byte` in a map, but not for `pKey`. Would be nice to avoid the allocation for `sKey` when possible, but I'm not seeing a clear way to restructure the code to do that right now, so perhaps just leave a TODO.